standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java

```diff
@@ -5269,6 +5293,32 @@ private void preDropStorageDescriptor(MStorageDescriptor msd) {
     removeUnusedColumnDescriptor(mcd);
   }
 
+  /**
+   * Get a list of storage descriptors that reference a particular Column Descriptor
+   * @param oldCD the column descriptor to get storage descriptors for
+   * @return a list of storage descriptors
+   */
+  private List<MStorageDescriptor> listStorageDescriptorsWithCD(MColumnDescriptor oldCD, Query query) {
+    boolean success = false;
+    List<MStorageDescriptor> sds = null;
+    try {
+      openTransaction();
```

Comment:
Why did we move the transaction here?
If the previous check runs in a different transaction then we might end up in a situation where the check did not found any usage for the descriptor but before the transaction for the removal started someone inserted new data and started to use the CD.

I think it is only theoretical possibility, but we should check this out.

Thanks,
Peter

Reply:
In my understanding transactions are usually used when one has CREATE, UPDATE or DELETE statements and one wants to have the atomic behavior, that is, either commit everything or commit nothing. Consider all queries we have here

1. Query for postgres

        SELECT * FROM "SDS" "A0" WHERE "A0"."CD_ID" = $1 limit 1)

2. Query for all DBs except postgres

        select count(1) from org.apache.hadoop.hive.metastore.model.MStorageDescriptor where (this.cd == inCD)

3. Constraint deletion

        pm.deletePersistentAll(mConstraintsList);

4. CD deletion

        pm.deletePersistent(oldCD);

IMHO we need transaction only for items 3 and 4 and it should look this way:

        1. SELECT * FROM "SDS" "A0" WHERE "A0"."CD_ID" = $1 limit 1)
        2. select count(1) from org.apache.hadoop.hive.metastore.model.MStorageDescriptor where (this.cd == inCD)
        openTransaction()
        3. pm.deletePersistentAll(mConstraintsList);
        4. pm.deletePersistent(oldCD);
        commitTransaction()

whilst the old implementation was

        openTransaction()        
        1. select count(1) from org.apache.hadoop.hive.metastore.model.MStorageDescriptor where (this.cd == inCD)        
        2. pm.deletePersistentAll(mConstraintsList);
        3. pm.deletePersistent(oldCD);
        commitTransaction()  

which seems to me redundant. Make sense?

Reply:
Transactions could be used also to prevent other queries modifying the data which was read by another transaction depending on the isolation level used (for SERIALIZABLE isolation level). [One good description](https://www.sqlservercentral.com/articles/isolation-levels-in-sql-server)

If there are concurrent transactions that could be an issue:
- T1 reads the data - decides that the CD is not used anymore
- T2 inserts a new row
- T1 drops the CD

The previous version of the code read the data inside the transaction, we can change the behavior if we decide so, but that has to be a conscious decision.

IMHO this is all theoretical since if I remember correctly the `openTransaction()` is called before this method call, so in our case the new `openTranscation()` call just increments a counter.

Reply:
Ok, I see. So your suggestion is to use transactions this way

    openTransaction()
    1. SELECT * FROM "SDS" "A0" WHERE "A0"."CD_ID" = $1 limit 1)
    2. select count(1) from org.apache.hadoop.hive.metastore.model.MStorageDescriptor where (this.cd == inCD)
    3. pm.deletePersistentAll(mConstraintsList);
    4. pm.deletePersistent(oldCD);
    commitTransaction()

to prevent any possible changes in tables we read. Correct? If so I will change the PR in this manner.

Reply:
And what do you think of method _listStorageDescriptorsWithCD_? It also uses open/commit transaction and If I implement open/commit transaction in _removeUnusedColumnDescriptor()_ it seems to be like this
       
       removeUnusedColumnDescriptor(){
         openTransaction()
         if (isPostgres()) {
           listStorageDescriptorsWithCD {  
             openTransaction()  <--- do we need this?
             1. SELECT * FROM "SDS" "A0" WHERE "A0"."CD_ID" = $1 limit 1)
             commitTransaction()   <--- do we need this?
             }
           }
         2. select count(1) from org.apache.hadoop.hive.metastore.model.MStorageDescriptor where (this.cd == inCD)
         3. pm.deletePersistentAll(mConstraintsList);
         4. pm.deletePersistent(oldCD);
         commitTransaction()
        }


so open/commit transaction  inside _listStorageDescriptorsWithCD()_ is redundant. Agree?

Reply:
We the  goal is with `openTransaction()` and the `commitTransaction()` to allow embedding method calls and keeping / enforcing transactions around them.

So if `listStorageDescriptorsWithCD()` is never called from non-transactional methods then adding `openTransaction()` and `commitTransaction()` is redundant. We wrap the calls into transactions if want to make sure to run the code in a transaction whenever it is called.

Wrapping a read only query in a transaction does not really make sense. It is either good without a transaction, or it requires a transaction because of the wider context, and then it is the callers responsibility to handle the transaction open/close. **So you are right, we do not need the transaction for the select itself.**